pulpcore/app/models/content.py

```diff
@@ -104,10 +105,44 @@ def touch(self):
         """
         Update the ``timestamp_of_interest`` on all objects of the query.
 
-        We order-by-pk here to avoid deadlocking in high-concurrency
-        environments.
-        """
-        return self.order_by("pk").update(timestamp_of_interest=now())
+        Postgres' UPDATE call doesn't support order-by. This can (and does) result in deadlocks in
+        high-concurrency environments, when using touch() on overlapping data sets. In order to
+        prevent this, we choose to SELECT FOR UPDATE with SKIP LOCKS == True, and only update
+        the rows that we were able to get locks on. Since a previously-locked-row implies
+        that updating that row's timestamp-of-interest is the responsibility of whoever currently
+        owns it, this results in correct data, while closing the window on deadlocks.
+        """
+        # Build the list of ids we need to work on, since we're going to be building a
+        # SQL-query "by hand" in a moment.
+        pulp_ids = [f"'{uuid}'" for uuid in self.values_list("pk", flat=True)]
+        if not pulp_ids:
+            return None
+        ids_str = ",".join(pulp_ids)
+        # timestamp_of_interest exists on core_content and core_artifact, not on the Detail tables
+        # If we have a master, we want to update the Content table, not the detail-table.
+        # If we do **not** have a master, use the table associated w/ the query.
+        db_table = (
+            Content._meta.db_table if hasattr(self.model, "master") else self.model._meta.db_table
```

Comment:
Sorry for not seeing the obvious earlier:
```suggestion
            Content._meta.db_table if isinstance(self, Content) else self.model._meta.db_table
```

`help(isinstance)`:
```
isinstance(obj, class_or_tuple, /)
    Return whether an object is an instance of a class or of a subclass thereof.
```

Reply:
...omg, i r dum. Great catch!

Reply:
To be fair, we talked about that very line and were both like "Hmm, ugly, but what else can we do?".

Reply:
Actually needed issubclass - we don't have an instance here, just the model class - but still a good improvement!